crates/loader/src/local/mod.rs

```diff
@@ -63,6 +63,20 @@ async fn prepare_any_version(
     }
 }
 
+/// Iterates over a vector of RawComponentManifest structs and return an error if any component ids are duplicated
+fn error_on_duplicate_ids(components: Vec<RawComponentManifest>) -> Result<()> {
+    let mut ids: Vec<String> = Vec::new();
+    for c in components {
+        let id = c.id;
+        if ids.contains(&id) {
+            return Err(anyhow!("cannot have duplicate component IDs: {}", id));
```

Comment:
nit: you can use `bail!` here as well.

Reply:
I had originally written it that way but was running into some issues

basically:
```rust
if ids.contains(&id) {
    bail!("cannot have duplicate component IDs: {}", id);
}

...

error_on_duplicate_ids(raw.components.clone()); // warning: unused `Result` that must be used
``` 
I got rid of the warning by writing it as
```rust
_ = error_on_duplicate_ids(raw.components.clone());
```

The test case also did not work as is currently written, the `from_file(...).await` call panicked, and the following `app.is_err()` and the "error contains `hello`" couldn't be reached.

My thinking was for the constraint function to just return a result and let the caller handle it, as opposed to panic'ing inside the constraint function.

I think the _real_ issue I'm having is understanding how to structure the test if the main function `prepare` isn't the one making the `bail!` call.